File: where_am_i/src/anomaly_detection.cpp
#include <ros/ros.h>
#include <sensor_msgs/Imu.h>
#include <eigen3/Eigen/Dense> 

#include <cstdlib>
#include <cmath>

class AnomalyDetector {
public:
    AnomalyDetector() {
        ros::NodeHandle nh;

        imu_subscriber = nh.subscribe("/imu", 1000, &AnomalyDetector::imuCallback, this);

        acceleration_threshold = 10.0; // Modify as needed acording to max values of your robot
        angular_velocity_threshold = 10.0; // Modify as needed acording to max values of your robot
        first_start = true;
    }

    void imuCallback(const sensor_msgs::Imu::ConstPtr& data) {
        double linear_acceleration = std::sqrt(std::pow(data->linear_acceleration.x, 2) +
                                               std::pow(data->linear_acceleration.y, 2) +
                                               std::pow(data->linear_acceleration.z, 2));

        double angular_velocity = std::sqrt(std::pow(data->angular_velocity.x, 2) +
                                            std::pow(data->angular_velocity.y, 2) +
                                            std::pow(data->angular_velocity.z, 2));

        if (linear_acceleration > acceleration_threshold ||
            angular_velocity > angular_velocity_threshold ||
            first_start) {
            ROS_INFO("Anomaly detected in IMU data! Activating kidnapper algorithm.");
            first_start = false;
            activateKidnapperAlgorithm();
        } else {
            ROS_INFO("IMU data within normal range.");
        }
    }

private:
    ros::Subscriber imu_subscriber;
    double acceleration_threshold;
    double angular_velocity_threshold;
    bool first_start;

    void activateKidnapperAlgorithm() {
        ROS_INFO("Activating kidnapper algorithm as a separate process.");
        std::system("/home/alp/catkin_ws/src/turtlebot3_kalman_filter/src/the_kidnapper_algorithm");
}

};
int main(int argc, char **argv) {
    ros::init(argc, argv, "anomaly_detector_node_cpp"); 

    AnomalyDetector anomalyDetector; 

    ros::spin();

    return 0;
}

File: where_am_i/src/anomaly_detection.py
#!/usr/bin/env python3

import rospy
from sensor_msgs.msg import Imu
import numpy as np
from The_kidnapper_Algorithm import *
from geometry_msgs.msg import Twist
import subprocess

class AnomalyDetector:
    def __init__(self):
        rospy.init_node('imu_anomaly_detector')
        self.First_start=True
        self.imu_subscriber = rospy.Subscriber('/imu', Imu, self.imu_callback)
        self.acceleration_threshold = 10
        self.angular_velocity_threshold = 1 

    def imu_callback(self, data):
        
        linear_acceleration = np.sqrt(data.linear_acceleration.x**2 +
                                      data.linear_acceleration.y**2 +
                                      data.linear_acceleration.z**2)

        angular_velocity = np.sqrt(data.angular_velocity.x**2 +
                                   data.angular_velocity.y**2 +
                                   data.angular_velocity.z**2)

        
        if (linear_acceleration > self.acceleration_threshold or
            angular_velocity > self.angular_velocity_threshold or self.First_start==True):
            rospy.loginfo("Anomaly detected in IMU data! Activating kidnapper algorithm.")
            pub = rospy.Publisher("/cmd_vel", Twist, queue_size=10)
            stop_movement = Twist()
            stop_movement.linear.x = 0
            stop_movement.angular.z = 0
            pub.publish(stop_movement)
            self.First_start=False
            self.activate_kidnapper_algorithm()
            
        else:
            rospy.loginfo("IMU data within normal range.")

    def activate_kidnapper_algorithm(self):
        rospy.loginfo("Activating kidnapper algorithm as a separate process.")
        subprocess.run(["python3", "/home/alp/catkin_ws/src/nerdeyim/turtlebot3/where_am_i/src/The_kidnapper_Algorithm.py"])  ####change it to where your python script is


def main():
    anomaly_detector = AnomalyDetector()
    rospy.spin()

if __name__ == '__main__':
    main()
